appmap/_implementation/recorder.py

```diff
@@ -1,81 +1,184 @@
 import logging
 import threading
 import traceback
+from abc import ABC, abstractmethod
 
-from .event import Event, _EventIds
-from .metadata import Metadata
+from appmap._implementation.utils import appmap_tls
 
 logger = logging.getLogger(__name__)
 
+_default_recorder = None
 
-class Recorder:
 
-    # _recorders is a dict of the in-progress recordings. It's keys are an thread-specific
-    # recorders, and None if a global recorder was created.
-    _recorders = None
-    _lock = threading.Lock()
+class Recorder(ABC):
+    """
+    A base class for Recorders.
+
+    Note that the abstract methods have implementations for use by subclasses.
+    """
+
+    @property
+    @abstractmethod
+    def events(self):
+        return self._events
+
+    @property
```

Comment:
I don't think this should be a `@property`; I wouldn't expect a property access to mutate state (except for caching or reification).

Reply:
Done.